src/audio/kpb.c

```diff
@@ -749,12 +752,69 @@ static void kpb_init_draining(struct comp_data *kpb, struct kpb_client *cli)
 	}
 }
 
+/**
+ * \brief Draining task.
+ *
+ * \param[in] arg - pointer keeping drainig data previously prepared
+ * by kpb_init_draining().
+ *
+ * \return none.
+ */
 static uint64_t kpb_draining_task(void *arg)
 {
-	/* TODO: while loop drainning history buffer accoriding to
-	 * clients request
-	 */
-	return 0;
+	struct dd *draining_data = (struct dd *)arg;
+	struct comp_buffer *sink = draining_data->sink;
+	struct hb *buff = draining_data->history_buffer;
+	size_t history_depth = draining_data->history_depth;
+	size_t size_to_read;
+	size_t size_to_copy;
+	uint32_t draining_done = 0;
+	uint32_t move_buffer = 0;
+	uint64_t ret = 1;
+
+	trace_kpb("kpb_draining_task(), start.");
+
+	while (history_depth > 0 && !draining_done) {
+		size_to_read = (uint32_t)buff->end_addr - (uint32_t)buff->r_ptr;
+
+		if (size_to_read > sink->free) {
+			if (sink->free >= history_depth) {
+				size_to_copy = history_depth;
+				draining_done = 1;
+			} else {
+				size_to_copy = sink->free;
+			}
+		} else {
+			if (size_to_read >= history_depth) {
+				size_to_copy = history_depth;
+				draining_done = 1;
+			} else {
+				size_to_copy = size_to_read;
+				move_buffer = 1;
+			}
+		}
+
+		memcpy(sink->w_ptr, buff->r_ptr, size_to_copy);
+		buff->r_ptr += (uint32_t)size_to_copy;
+		history_depth -= size_to_copy;
+
+		if (move_buffer) {
+			buff->r_ptr = buff->start_addr;
+			buff = buff->next;
+			move_buffer = 0;
+		}
+		if (size_to_copy)
+			comp_update_buffer_produce(sink, size_to_copy);
+	}
+
+	if (draining_done) {
+		/* switch KPB to copy real time stream to client's sink */
+		*draining_data->state = KPB_DRAINING_ON_DEMAND;
```

Comment:
Please check the latest changes regarding hda blocking copy.
You can switch to blocking mode in kpb_prepare() and switch back here:
`comp_set_attribute(sink->sink, COMP_ATTR_COPY_BLOCKING, 0)`

Reply:
We need hda-dma blocking mode only for draining, so I think we should switch to blocking bode on init_draining and reset to normal mode here, in 

> if (draining_done)